Failed to post new order: operation timed out

Just happened on regtest. The post order http request timed out. 

```
[DEBUG] TIME: 2024-02-17T06:37:51.714612Z r: Pruning expired orders 
[ERROR] TIME: 2024-02-17T06:37:51.714855Z r: Failed to post new order: error sending request for url (http://34.32.0.52/api/orderbook/orders): operation timed out: operation timed out order_id: 58c11fda-d6ea-4129-a962-875098d5ab65
[DEBUG] TIME: 2024-02-17T06:37:51.715370Z r: lightning::ln::channelmanager: Marking channels with 03507b924dae6595cfb78492489978127c5f1e3877848564de2015cd6d41375802 disconnected and generating channel_updates. 
[DEBUG] TIME: 2024-02-17T06:37:51.715581Z r: Connection lost; reconnecting peer: 03507b924dae6595cfb78492489978127c5f1e3877848564de2015cd6d41375802@34.32.0.52:9045,reconnect_interval: 1s
[INFO] TIME: 2024-02-17T06:37:51.715634Z r: Updated order state new_state: Failed,order_id: 58c11fda-d6ea-4129-a962-875098d5ab65
[ERROR] TIME: 2024-02-17T06:37:51.716063Z Failed to submit order: FfiException(RESULT_ERROR, Failed to post order to orderbook: error sending request for url (http://34.32.0.52/api/orderbook/orders): operation timed out, null)
```

Not sure if this ERROR is part of this bug or something else. However it happened in the same milli second as the post order request timed out.

```
[ERROR] TIME: 2024-02-17T06:37:51.713745Z r: Failed to handle event: Trade request sent to coordinator for order 58c11fda-d6ea-4129-a962-875098d5ab65 failed: Could not load order from db: cannot acquire database connection: timed out waiting for connection 
```

The good news is that this bug doesn't seem to block the user a lot. After the order times out (1 minute). The user is able to trade again. I was able to retry successfully.

I managed to reproduce this issue by adding a timeout here.

```
#[instrument(skip_all, err(Debug))]
pub async fn post_order(
    State(state): State<Arc<AppState>>,
    Json(new_order): Json<NewOrder>,
) -> Result<Json<Order>, AppError> {
    let (sender, mut receiver) = mpsc::channel::<Result<Order>>(1);

    let message = NewOrderMessage {
        new_order,
        order_reason: OrderReason::Manual,
        sender,
    };
    state.trading_sender.send(message).await.map_err(|e| {
        AppError::InternalServerError(format!("Failed to send new order message: {e:#}"))
    })?;

    thread::sleep(Duration::from_secs(10));

    let result = receiver
        .recv()
        .await
        .context("Failed to receive response from trading sender")
        .map_err(|e| AppError::InternalServerError(format!("{e:#}")))?;

    let order = result.map_err(|e| match e.downcast_ref() {
        Some(TradingError::InvalidOrder(reason)) => AppError::InvalidOrder(reason.to_string()),
        Some(TradingError::NoMatchFound(message)) => AppError::NoMatchFound(message.to_string()),
        _ => AppError::InternalServerError(format!("Failed to post order. Error: {e:#}")),
    })?;

    Ok(Json(order))
}
```

My theory is that it could happen that the the order is processed before we even start to listen on the receiver. If that happens the http request would never return because the `receiver.recv().await` will never finish.

Adding the 10 sec sleep in between forced that behaviour to happen and resulted into the exact same error logs on the app side like shown above and reported in the support ticket. 

see below

```
flutter: [INFO] TIME: 2024-02-17T08:44:24.145138Z r: Received match from orderbook order_id: 12a540c9-83f2-4f40-8542-75a9d1ad0507
flutter: [ERROR] TIME: 2024-02-17T08:44:54.144741Z r: Failed to handle event: Trade request sent to coordinator for order 12a540c9-83f2-4f40-8542-75a9d1ad0507 failed: Could not load order from db: cannot acquire database connection: timed out waiting for connection
flutter: [DEBUG] TIME: 2024-02-17T08:44:54.145207Z r: Pruning expired orders
flutter: [TRACE] TIME: 2024-02-17T08:44:54.145376Z r: lightning_background_processor: Calling PeerManager's timer_tick_occurred
flutter: [ERROR] TIME: 2024-02-17T08:44:54.145560Z r: Failed to post new order: error sending request for url (http://127.0.0.1:8000/api/orderbook/orders): operation timed out: operation timed out order_id: 12a540c9-83f2-4f40-8542-75a9d1ad0507
flutter: [TRACE] TIME: 2024-02-17T08:44:54.145675Z r: lightning_background_processor: Rebroadcasting monitor's pending claims
flutter: [INFO] TIME: 2024-02-17T08:44:54.145791Z r: Updated order state new_state: Failed,order_id: 12a540c9-83f2-4f40-8542-75a9d1ad0507
flutter: [ERROR] TIME: 2024-02-17T08:44:54.156429Z Failed to submit order: FfiException(RESULT_ERROR, Failed to post order to orderbook: error sending request for url (http://127.0.0.1:8000/api/orderbook/orders): operation timed out, null)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Fix(mobile): Calculate counterparty balance correctly when checking the validity of trade parameters.
 - Fix(trade): Spawn dedicated tokio task when executing trade.
 - Chore(wallet): Unreserve locked utxos if offer get rejected.
+- Fix(order): Insert channel-opening parameters before posting order to the coordinator. This prevents the app from running into a race condition between the completion of the HTTP request and the app being notified about the order being matched.
 
 ## [1.8.8] - 2024-02-14
 

--- a/mobile/lib/features/trade/submit_order_change_notifier.dart
+++ b/mobile/lib/features/trade/submit_order_change_notifier.dart
@@ -94,23 +94,21 @@ class SubmitOrderChangeNotifier extends ChangeNotifier implements Subscriber {
     if (event is bridge.Event_OrderUpdateNotification) {
       Order order = Order.fromApi(event.field0);
 
-      if (_pendingOrder?.id == order.id) {
-        switch (order.state) {
-          case OrderState.open:
-          case OrderState.filling:
-            return;
-          case OrderState.filled:
-            _pendingOrder!.state = PendingOrderState.orderFilled;
-            break;
-          case OrderState.failed:
-          case OrderState.rejected:
-            _pendingOrder!.state = PendingOrderState.orderFailed;
-            break;
-        }
-        _pendingOrder!.failureReason = order.failureReason;
-
-        notifyListeners();
+      switch (order.state) {
+        case OrderState.open:
+        case OrderState.filling:
+          return;
+        case OrderState.filled:
+          _pendingOrder!.state = PendingOrderState.orderFilled;
+          break;
+        case OrderState.failed:
+        case OrderState.rejected:
+          _pendingOrder!.state = PendingOrderState.orderFailed;
+          break;
       }
+      _pendingOrder!.failureReason = order.failureReason;
+
+      notifyListeners();
     } else {
       logger.w("Received unexpected event: ${event.toString()}");
     }

--- a/mobile/native/src/db/models.rs
+++ b/mobile/native/src/db/models.rs
@@ -1477,14 +1477,16 @@ impl ChannelOpeningParams {
     pub fn insert(
         conn: &mut SqliteConnection,
         channel_opening_params: ChannelOpeningParams,
-    ) -> Result<()> {
+    ) -> QueryResult<()> {
         let affected_rows = diesel::insert_into(channel_opening_params::table)
             .values(channel_opening_params)
             .execute(conn)?;
 
-        ensure!(affected_rows > 0, "Could not insert channel-opening params");
+        if affected_rows == 0 {
+            return diesel::result::QueryResult::Err(diesel::result::Error::NotFound);
+        }
 
-        Ok(())
+        diesel::result::QueryResult::Ok(())
     }
 
     pub fn get_by_order_id(

--- a/mobile/native/src/trade/order/handler.rs
+++ b/mobile/native/src/trade/order/handler.rs
@@ -18,7 +18,6 @@ use anyhow::anyhow;
 use anyhow::bail;
 use anyhow::Context;
 use anyhow::Result;
-use diesel::Connection;
 use reqwest::Url;
 use time::Duration;
 use time::OffsetDateTime;
@@ -84,40 +83,32 @@ pub async fn submit_order(
         });
     }
 
+    db::insert_order(order.clone()).map_err(SubmitOrderError::Storage)?;
+
+    if let Some(channel_opening_params) = channel_opening_params {
+        let runtime = get_or_create_tokio_runtime().expect("to be able to get runtime");
+        runtime
+            .spawn_blocking({
+                move || {
+                    let mut conn = db::connection()?;
+                    tracing::debug!(
+                        ?channel_opening_params,
+                        "Recording channel-opening parameters"
+                    );
+
+                    db::insert_channel_opening_params(&mut conn, channel_opening_params)
+                }
+            })
+            .await
+            .expect("task to complete")
+            .map_err(SubmitOrderError::Storage)?;
+    }
+
     let url = format!("http://{}", config::get_http_endpoint());
     let url = Url::parse(&url).expect("correct URL");
     let orderbook_client = OrderbookClient::new(url);
 
-    db::insert_order(order.clone()).map_err(SubmitOrderError::Storage)?;
-
-    let runtime = get_or_create_tokio_runtime().expect("to be ablet to get runtime");
-    if let Err(err) = runtime
-        .spawn_blocking({
-            let order = order.clone();
-            move || {
-                let mut db = db::connection()?;
-                // We need to ensure that the channel-opening parameters are inserted after the
-                // order has been posted, but before order has been match and we need them to open
-                // the DLC channel. Without this database transaction we can easily end up with a
-                // match before we insert them!
-                db.transaction(|conn| {
-                    if let Some(channel_opening_params) = channel_opening_params {
-                        tracing::debug!(
-                            ?channel_opening_params,
-                            "Recording channel-opening parameters"
-                        );
-
-                        db::insert_channel_opening_params(conn, channel_opening_params)
-                            .map_err(SubmitOrderError::Storage)?;
-                    }
-
-                    runtime.block_on(orderbook_client.post_new_order(order.into()))
-                })
-            }
-        })
-        .await
-        .expect("task to complete")
-    {
+    if let Err(err) = orderbook_client.post_new_order(order.clone().into()).await {
         let order_id = order.id.clone().to_string();
 
         tracing::error!(order_id, "Failed to post new order: {err:#}");

